plugins/statuslines/lualine.nix

```diff
@@ -1,260 +1,281 @@
+      left = defaultNullOpts.mkStr leftDefault "Left separator";
+      right = defaultNullOpts.mkStr rightDefault "Right separator";
     };
 
   mkComponentOptions =
     defaultName:
-    helpers.mkNullOrOption (
-      with types;
-      listOf (
-        either str (submodule {
-          options = {
-            name = mkOption {
-              type = types.either types.str helpers.nixvimTypes.rawLua;
-              description = "Component name or function";
-              default = defaultName;
-            };
-
-            icons_enabled = helpers.defaultNullOpts.mkBool true ''
-              Enables the display of icons alongside the component.
-            '';
-
-            icon = helpers.mkNullOrOption (
-              with types;
-              either str (submodule {
+    lib.nixvim.mkNullOrOption' {
+      type =
+        with types;
+        let
+          # TODO: added 2024-09-05 remove after 24.11
+          oldAttrs = [
+            "name"
+            "icon"
+            "extraConfig"
+          ];
+          isOldType = def: lib.any (name: lib.elem name oldAttrs) (lib.attrNames def);
+          oldType = addCheck (attrsOf anything) isOldType // {
+            description = "attribute set containing ${lib.concatStringsSep ", " oldAttrs}";
+          };
+          coerceFn =
+            option:
+            if option != null && lib.isAttrs option then
+              let
+                # NOTE: Handle deprecated name property
+                # This was used to create an unkeyed attribute for a component name
+                optionWithNameTransformed =
+                  if lib.hasAttr "name" option then
+                    lib.removeAttrs (option // { "__unkeyed" = option.name; }) [ "name" ]
+                  else
+                    option;
+
+                # NOTE: Handle deprecated nested icon option
+                # This was used to create an unkeyed attribute for an icon with options
+                optionWithIconTransformed =
+                  if
+                    lib.hasAttr "icon" optionWithNameTransformed
+                    && lib.isAttrs optionWithNameTransformed.icon
+                    && lib.hasAttr "icon" optionWithNameTransformed.icon
+                  then
+                    lib.removeAttrs (
+                      optionWithNameTransformed.icon // { "__unkeyed" = optionWithNameTransformed.icon.icon; }
+                    ) [ "icon" ]
+                  else
+                    optionWithNameTransformed;
+
+                # NOTE: Handles the deprecated extraConfig option
+                # Old freeform option was just merged into the existing config
+                flattenedExtraConfig = optionWithIconTransformed.extraConfig or { };
+                optionWithoutExtraConfig = lib.removeAttrs optionWithIconTransformed [ "extraConfig" ];
+                finalTransformedValue = optionWithoutExtraConfig // flattenedExtraConfig;
```

Comment:
Let's pipe this:

First, replace `optionsWithXYZTransformed` with various coercion functions, `transformName`, `transformIcon`, etc.

Then we can `pipe` the options attr through those transformers:

```nix
coerceFn = attrs: lib.pipe attrs [
  # Transform old `name` attr to `__unkeyed`
  (
    x:
      if x ? name then
        lib.removeAttrs option [ "name" ] // {
          __unkeyed-1 = option.name;
        }
      else
       x
  )
  # Transform old `icon.icon` attr to `__unkeyed`
  # FIXME: this conflicts with `name`, is that intended?
  (
    x:
      if x.icon or null ? icon then
        removeAttrs x [ "icon" ] // {
          __unkeyed-1 = x.icon.icon;
        }
      else
        x
  )
  # Merge in old `extraConfig` attr
  (x: removeAttrs x [ "extraConfig" ] // x.extraConfig or { })
];
```

TBH, I don't think it's less readable with the transformers as inline anonymous lambdas, as above...

Reply:
>   # Transform old `icon.icon` attr to `__unkeyed`
>   # FIXME: this conflicts with `name`, is that intended?
>   (
>     x:
>       if x.icon or null ? icon then
>         removeAttrs x [ "icon" ] // {
>           __unkeyed-1 = x.icon.icon;
>         }
>       else
>         x
>   )

What do you mean about the conflict? There is a transformation we did for nested icon attributes that behaved similar to the name attribute. Basically, we had a icon = string or a icon attrset with an icon = string gets changed to unkeyed.

Reply:
I'm not too familiar, but just reading the transformers above, both `icon.icon` _and_` `name` are converted into `__unkeyed-1`.

Currently, if both `name` and `icon.icon` are defined, `icon.icon` will win.

Reply:
There *shouldn't*  have been a usage for name within icon, before... so I'm not too worried about breaking that config.

Reply:
Ok, it seems the icon transformer should be:

```nix
(
  x:
    if isAttrs (x.icon or null) && x.icon ? icon then
      x // { icon = removeAttrs x.icon [ "icon" ] // {
        __unkeyed-1 = x.icon.icon;
      }
    else
      x
)
```